posts/2020-02-12-using-dotty-union-types-with-akka-typed-part-II.adoc

```diff
@@ -0,0 +1,285 @@
+
+Does this make sense intuitively? It does: `externalBehavior` is declared as a behavior that is able to 'process' all messages that are part of the `Pinger` actor's `Command` protocol. The behavior that is actually handling these messages is the `internalBehavior` which, on top of the messages of type `Command`, is able to process the `PingPong.Pong` message.
+
+NOTE: One could say that from the outside, the internal behavior is not utilised to its full extent.
+
+The combination of Dotty Union types combined with generic `Behavior` having contravariant type parameter leads to a very simple implementation of the `Pinger` actor. The future will tell if, with Dotty, Akka will be able to exploit this in a future version.
+
+== Returning to best practices for coding up actors in Akka Typed
+
+I prefer to put the core behavior of a typed actor in a companion class. In simple cases, that may be considered as overkill, but as a counter argument, I think that applying the same practice in a consistent manner, it helps to maintain a recurring and easily recognisable way of coding actors.
+
+The Dotty version of the `Pinger` actor will then look as follows:
+
+[source,scala,linenums]
+----
+import akka.actor.typed.{ActorRef, Behavior}
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+
+object Pinger {
+
+  // My protocol
+  sealed trait Command
+  case object SendPing extends Command
+  case object StopPingPong extends Command
+
+  // My protocol + the responses I need to understand...
+  private type CommandsAndResponses = Command | PingPong.Response
+
+  def apply(pingPong: ActorRef[PingPong.Ping]): Behavior[Command] = {
+    val internalBehavior = Behaviors.setup[CommandsAndResponses] { context =>
+      (new Pinger(context, pingPong)).run()
+    }
+    internalBehavior.narrow
+  }
+}
+
+class Pinger private (context: ActorContext[Pinger.CommandsAndResponses], pingPong: ActorRef[PingPong.Ping]) {
+
+  import Pinger._
+
+  def run(): Behavior[CommandsAndResponses] =
+    Behaviors.receiveMessage {
+      case StopPingPong =>
+        context.log.info(s"End of the ping-pong game - calling it a day!")
+        context.system.terminate()
+        Behaviors.stopped
+      case SendPing =>
+        pingPong ! PingPong.Ping(replyTo = context.self)
+        Behaviors.same
+      case response : PingPong.Response =>
+        context.log.info(s"Hey: I just received a $response !!!")
+        Behaviors.same
+    }
+}
+----
+
+Note that we prevent the direct creation of instances of the `Pinger` actor by marking the constructor of the `Pinger` class private.
+
+=== Conclusions
+
+In this article we have:
+
+* Shown in both Scala 2 and Dotty, we can hide the handling of responses to messages sent to other actors:
+** In Scala 2 we can use message adapters.
+** In Dotty, we can use the `narrow` method on `Behavior`.
```

Comment:
```suggestion
** In Dotty we can use the `narrow` method on `Behavior`.
```

Reply:
A comma is fine here as well, but you didn't use one in the line above. So either use one in both, or don't use one in either.

Reply:
> A comma is fine here as well, but you didn't use one in the line above. So either use one in both, or don't use one in either.

I'll add the comma on the first line too.